### ISSUE TITLE
Difficult or impossible to reliably detect whether certain video encoding types are supported (e.g. AV1)

### ISSUE THREAD

**Issue by benstevens48** [microsoft/WindowsAppSDK#4476]
### Describe the bug

I am trying to determine whether video encoding using AV1 is supported. To do this I check ApiInformation.IsMethodPresent("Windows.Media.MediaProperties.MediaEncodingProfile", "CreateAv1"). However, this returns true on my machine, even though it says in the docs "Windows 11 Insider Preview (introduced in 10.0.23504.0)" and the method is not available when compiling (I am using the 22621 Windows SDK, not any insider preview SDK) and my machine is running 23H2. It's annoying because if I do try to encode AV1 video (not using this method but by using the "Av1" subtype string), it hangs indefinitely. I am struggling to think of how to limit the functionality. Maybe I will have to do a manual Windows version check, but I don't really want to do that.

### Steps to reproduce the bug

See description

### Expected behavior

ApiInformation.IsMethodPresent (and other related functions) should return false if the method is not really available. If the method is experimental or in preview, then there should be some way to determine that (and I can't see it, but please correct me if I'm wrong).

### NuGet package version

None

### Packaging type

Packaged (MSIX)

### Windows version

Windows 11 version 22H2 (22621, 2022 Update)

### IDE

Visual Studio 2022

### Additional context

Does not require WindAppSDK, is a WinRT issue.

**Comment by DarranRowe** [comment]
Are you sure that this is a bug and not the fact that these features could have been backported to this version of Windows 11?
Try building against the new Windows SDK (26100) and actually use CreateAv1.

**Comment by bencambridgemaths** [comment]
I just installed the new 26100 SDK, used MediaEncodingProfile.CreateAv1, and it still hangs in the same way. It really should never hang so there's a bug there anyway. 

Maybe the bug is that it just doesn't work.

Alternatively, it's not supposed to work currently, in which case there ought to be a way to tell that. It's possible that in fact the encoding capability is going to be added to an update to the AV1 Video Extensions package, which may be released at the same time as the 24H2 update, and then it will work down-level, but in the meantime how can we tell whether it's supported?

**Comment by DarranRowe** [comment]
One of the big issues here is:

![Screenshot 2024-06-13 185629](https://github.com/microsoft/WindowsAppSDK/assets/52577874/1928a096-8204-492b-b213-47e25075e7ae)

I don't see anything that you mention happening. My GPU (Radeon 7800 XT) does support hardware AV1 encoding though.

Out of curiosity, if you query for your encoders using [Windows.Media.Core.CodecQuery](https://learn.microsoft.com/en-us/uwp/api/windows.media.core.codecquery?view=winrt-22621), does that show an available AV1 encoder? I have one:

```
Encoder name: AMDh264Encoder

Encoder name: AMDh265Encoder

Encoder name: AMDh264Encoder

Encoder name: AMDav1Encoder <-- Here

Encoder name: AMDh265Encoder

Encoder name: HEIFImageExtension

Encoder name: VP9VideoExtensionEncoder

Encoder name: HEVCVideoExtensionEncoder

Encoder name: HEVCVideoExtensionEncoder

Encoder name: H264 Encoder MFT

Encoder name: WMVideo8 Encoder MFT

Encoder name: H263 Encoder MFT

Encoder name: WMVideo9 Encoder MFT

Encoder name: Microsoft MPEG-2 Video Encoder MFT
```

As I stated, my system does have hardware AV1 encoding.

**Comment by bencambridgemaths** [comment]
When I say it hangs, I mean encoding hangs i.e. 
```C#
  var encodingProfile = MediaEncodingProfile.CreateAv1(VideoEncodingQuality.HD720p);
  var preparedTranscoder = await transcoder.PrepareStreamTranscodeAsync(inputStream, outputStream, encodingProfile);
  if (!preparedTranscoder.CanTranscode) {
      return;
  }
 await preparedTranscoder.TranscodeAsync(); //Hangs here
```
Yes, apparently I do have an hardware encoder - Intel® Hardware Accelerated AV1 Encoder MFT.

**Comment by DarranRowe** [comment]
That means that the bug isn't in the AV1 functionality exposed by WinRT.
Now, there are two major questions, first how large is the file that you are transcoding? Transcoding isn't a fast process, and on my system a file that I was using for testing does take around a minute.

Another thing to remember is that TranscodeAsync returns an IAsyncActionWithProgress. This means that you can hook a progress routine up to it and have it report progress. Have you done that?

An example of this is:
```
    winrt::Windows::Foundation::IAsyncAction MainWindow::Click_Click(winrt::IInspectable const &, winrt::RoutedEventArgs const &)
    {
        auto profile = Windows::Media::MediaProperties::MediaEncodingProfile::CreateAv1(Windows::Media::MediaProperties::VideoEncodingQuality::HD720p);
        auto in_file = co_await Windows::Storage::StorageFile::GetFileFromPathAsync(LR"(C:\Users\Darran\Documents\wpin.mkv)");
        auto out_folder = co_await Windows::Storage::StorageFolder::GetFolderFromPathAsync(LR"(C:\Users\Darran\Documents)");
        Windows::Storage::StorageFile out_file{ nullptr };
        auto out_file_list = co_await out_folder.GetFilesAsync();
        for (auto file : out_file_list)
        {
            if (file.Name() == L"wpout.mp4")
            {
                out_file = file;
            }
        }

        if (out_file == nullptr)
        {
            out_file = co_await out_folder.CreateFileAsync(L"wpout.mp4");
        }

        Windows::Media::Transcoding::MediaTranscoder transcoder{};
        auto prep_transcoder = co_await transcoder.PrepareFileTranscodeAsync(in_file, out_file, profile);
        if (prep_transcoder.CanTranscode())
        {
            //Note that there is no co_await here.
            auto transcode_async = prep_transcoder.TranscodeAsync();
            //Add a callback to the Progress member, which is IAsyncActionWithProgress<double>.Progress.
            transcode_async.Progress([this](const Windows::Foundation::IAsyncActionWithProgress<double> &async_info, const double &progress)
                {
                    DispatcherQueue().TryEnqueue([this, progress]() {
                        auto fmt_string = std::format(L"{}", progress);
                        Content().Text(fmt_string);
                        }
                    );
                });
            //Now we co_await the async transcoder.
            co_await transcode_async;
            
        }

        Content().Text(L"File encode complete");

        co_return;
    }
```

It is C++ due to that being more comfortable for me, but this basically hooks up a progress callback, and this sets the contents of a TextBox to the stringified version of that progress.

![Screenshot 2024-06-14 190317](https://github.com/microsoft/WindowsAppSDK/assets/52577874/7dbbdac0-400a-46d8-b066-6b3adcb6a388)

As you can see, it is progressing. Finally, after it does report the transcode as complete:

![Screenshot 2024-06-14 190814](https://github.com/microsoft/WindowsAppSDK/assets/52577874/2d479834-f8f4-4d74-8abc-9601c13a9481)

The out file has a pretty substantial file size. The video track is around 20 minutes and CBR, so that is what I would expect.

**Comment by benstevens48** [comment]
The video is only small and takes a few seconds to encode to H.264 or H.265. I have been using the progress. Some interesting notes on this.
- MPEG-4 container, VideoProcessingAlgorithm Default, before I updated graphics drivers: get stuck after 1 progress callback, but is cancellable. 
- MPEG-4 container, VideoProcessingAlgorithm Default, after I updated graphics drivers: gets almost to the end then errors. 
- MPEG-4 container, VideoProcessingAlgorithm MrfCrf444 (forces software encoding apparently) : gets almost to the end then errors.
- ASF container, VideoProcessingAlgorithm Default, before I updated graphics drivers: get stuck after 1 progress callback, but is cancellable. 
- ASF container, VideoProcessingAlgorithm Default, after I updated graphics drivers: succeeds, but output file is not playable. 
- ASF container, VideoProcessingAlgorithm MrfCrf444 (forces software encoding apparently) : succeeds, but output file is not playable.

Also, on a related note, while testing supported container and encoding types, I found that trying to use NV12 with a MPEG-4 container (not designed to be supported I guess) will freeze the app.

In conclusion, it seems like AV1 video encoding is not really currently supported, at least on my machine. Probably a software encoder is needed (i.e. an updated AV1 Video Extension package that supports encoding).

I thought that maybe this was the reason that the API was not available in an earlier SDK, but I guess perhaps the IsMethodPresent etc check will never be enough for this sort of scenario so I am going to change the title to reflect how to reliably test for encoding support. Meanwile, in my app I will just mark AV1 encoding as experimental.

**Comment by codendone** [comment]
Logged an [internal bug](http://task.ms/52146955) for the media team. Closing as External.
